extensions/control-plane/provision/gcs-provision/src/main/java/org/eclipse/dataspaceconnector/gcp/storage/provision/GcsProvisioner.java

```diff
@@ -0,0 +1,125 @@
+        this.monitor = monitor;
+        this.storageService = storageService;
+        this.iamService = iamService;
+    }
+
+    @Override
+    public boolean canProvision(ResourceDefinition resourceDefinition) {
+        return resourceDefinition instanceof GcsResourceDefinition;
+    }
+
+    @Override
+    public boolean canDeprovision(ProvisionedResource resourceDefinition) {
+        return resourceDefinition instanceof GcsProvisionedResource;
+    }
+
+    @Override
+    public CompletableFuture<StatusResult<ProvisionResponse>> provision(
+            GcsResourceDefinition resourceDefinition, Policy policy) {
+        String bucketName = resourceDefinition.getId();
+        String bucketLocation = resourceDefinition.getLocation();
+
+        monitor.debug("GCS Bucket request submitted: " + bucketName);
+
+        String resourceName = resourceDefinition.getId() + "-bucket";
+        try {
+            BucketWrapper bucket = storageService.getOrCreateBucket(bucketName, bucketLocation);
+            if (!storageService.isEmpty(bucketName)) {
+                return completedFuture(StatusResult.failure(ResponseStatus.FATAL_ERROR, String.format("Bucket: %s already exists and is not empty.", bucketName)));
+            }
+            ServiceAccountWrapper serviceAccount = createServiceAccount();
+            GcsAccessToken token = createBucketAccessToken(bucket, serviceAccount);
+
+            GcsProvisionedResource resource = getProvisionedResource(resourceDefinition, resourceName, serviceAccount);
+
+            var response = ProvisionResponse.Builder.newInstance().resource(resource).secretToken(token).build();
+            return CompletableFuture.completedFuture(StatusResult.success(response));
+        } catch (GcpExtensionException e) {
+            return completedFuture(StatusResult.failure(ResponseStatus.FATAL_ERROR, e.toString()));
+        }
+    }
+
+    @Override
+    public CompletableFuture<StatusResult<DeprovisionedResource>> deprovision(
+            GcsProvisionedResource provisionedResource, Policy policy) {
+        try {
+            iamService.deleteServiceAccountIfExists(
+                    new ServiceAccountWrapper(provisionedResource.getServiceAccountEmail(),
+                            provisionedResource.getServiceAccountName()));
+        } catch (GcpExtensionException e) {
+            return completedFuture(StatusResult.failure(ResponseStatus.FATAL_ERROR,
+                    String.format("Deprovision failed with: %s", e.getMessage())));
+        }
+        return CompletableFuture.completedFuture(StatusResult.success(
+                DeprovisionedResource.Builder.newInstance()
+                        .provisionedResourceId(provisionedResource.getId()).build()));
+    }
+
+    private ServiceAccountWrapper createServiceAccount() {
+        String uniqueId = UUID.randomUUID().toString().replace("-", "").substring(0, 26);
```

Comment:
nit: this could be put in something like a `sanitize` method. just a suggestion though.

Reply:
Done. Also made the provisioning idempotent by basing the name on the transfer process id